projects/hip/docs/how-to/performance_guidelines.rst

```diff
@@ -1,321 +1,623 @@
+
+.. code-block:: cuda
+
+   // Use naturally aligned types
+   float4 data;  // 16-byte aligned
+   float2 data;  // 8-byte aligned
+   
+   // Ensure structure alignment
+   struct __align__(16) MyStruct {
+       float4 data;
+   };
+
+**Optimize 2D array access**
+
+Padding 2D arrays to multiples of the wavefront size ensures each row starts 
+at an aligned memory boundary. This allows consecutive threads accessing the 
+same row to generate coalesced memory transactions, maximizing bandwidth.
 
-Optimization for maximum instruction throughput
-================================================================================
+.. code-block:: cuda
 
-To maximize instruction throughput:
+   // Ensure array width is multiple of warp size
+   int width = ((actual_width + warpSize - 1) / warpSize) * warpSize;
+   hipMalloc(&array, width * height * sizeof(float));
+   
+   // Access pattern
+   int idx = x + width * y;  // width should be warp-aligned
 
-- Minimize low throughput arithmetic instructions.
-- Minimize divergent warps inflicted by flow control instructions.
-- Maximize instruction parallelism.
+**Coalesce memory accesses**
 
-These techniques are discussed in detail in the following sections.
+When consecutive threads in a wavefront access consecutive memory addresses, 
+the hardware combines these into a single wide transaction. Non-coalesced 
+patterns require multiple transactions, reducing effective bandwidth.
+
+.. code-block:: cuda
+
+   // Good: consecutive threads access consecutive addresses
+   int idx = threadIdx.x + blockIdx.x * blockDim.x;
+   data[idx] = value;
+   
+   // Bad: strided access
+   int idx = threadIdx.x * stride;  // Non-coalesced if stride > 1
+   data[idx] = value;
+
+For understanding memory coalescing theory, see :ref:`memory_hierarchy_theory`.
+
+**Use shared memory for data reuse**
+
+Shared memory (LDS) provides low-latency on-chip storage shared across threads 
+in a block. Loading data into shared memory once and reusing it many times 
+reduces global memory traffic, particularly effective for tiled algorithms like matrix multiplication.
+
+.. code-block:: cuda
+
+   __global__ void optimized_kernel(float* input, float* output) {
+       __shared__ float tile[TILE_SIZE][TILE_SIZE];
+       
+       // Load data into shared memory
+       tile[threadIdx.y][threadIdx.x] = input[...];
+       __syncthreads();
+       
+       // Reuse data from fast shared memory
+       float result = 0;
+       for (int i = 0; i < TILE_SIZE; i++) {
+           result += tile[threadIdx.y][i] * tile[i][threadIdx.x];
+       }
+       __syncthreads();
+       
+       output[...] = result;
+   }
+
+**Avoid bank conflicts in shared memory**
+
+Shared memory is organized into banks that can service one request per cycle. 
+When multiple threads in a warp access the same bank simultaneously, they 
+serialize, reducing throughput. Padding arrays by one element shifts addresses 
+to avoid systematic conflicts.
```

Comment:
```suggestion
Shared memory is organized into banks, each capable of servicing one request per cycle. When multiple threads in a warp access the same bank simultaneously, the requests are serialized, reducing throughput. Padding arrays by one element shifts addresses to avoid systematic conflicts.
```